### DISCUSSION TITLE
Questions on nifti data loading with GPU

### DISCUSSION THREAD

**Discussion by yiheng-wang-nv** [nipy/nibabel#1385] [Q&A] [answered]
Hi Nibabel experts,

recently I did some research on how to use GPU Direct Storage (GDS) for NIFTI data loading, and I have some successful trials on Decathlon datasets.
My draft implementation is in: https://github.com/Project-MONAI/MONAI/pull/8188 and it uses `Nibabel + kvikio + cupy` for the loading.
Generally, I use `cupy` to create an empty array , and use `kvikio` to read the data into it:

```
        file_size = os.path.getsize(filename)
        image = cp.empty(file_size, dtype=cp.uint8)
        with kvikio.CuFile(filename, "r") as f:
            f.read(image)
```
Then, I use the following code to get the header, and the fdata:
```
        header_bytes = cp.asnumpy(image[:348])
        header = nib.Nifti1Header.from_fileobj(io.BytesIO(header_bytes))
        data_offset = header.get_data_offset()
        data_shape = header.get_data_shape()
        data_dtype = header.get_data_dtype()
        affine = header.get_best_affine()
        data = image[data_offset:].view(data_dtype).reshape(data_shape, order="F")

```
I have some questions on the above implementation:
1. Does Nibabel already have GDS related feature? If so, I can use it directly
2. Does my way of loading header and using offset, dtype and shape have any issues in some cases?

It would be very appreciate if you can help to answer them, thanks very much!

**Comment by effigies** [comment, marked as answer]
1. No.
2. I don't know if `kvikio.CuFile` handles compression, but many NIfTI files are gzipped. You will need to account for that if it's not done automatically. The construction of the header, affine and data seem reasonable. However, if there are scale factors, you will not get the same data. Also, nibabel lazily loads data, so you can get the header and your parameters like:

```python
img = nb.load(filename)
shape = img.shape
offset = img.dataobj.offset
data_dtype = img.dataobj.dtype
slope = img.dataobj.slope
inter = img.dataobj.inter

raw = ...  # Do what's needed to get your cupy array.
if (slope, inter) != (1, 0):
    data = slope * raw + inter
else:
    data = raw
```

**Reply by yiheng-wang-nv** [reply]
Thanks @effigies for the answer.
You are correct, for .nii.gz files, I did not figure out a fully GPU way to handle it.
Thanks for the explain of lazily loading, I will enhance it accordingly!
Thanks for the explain of scale factors, will include it in my formal implementation

**Reply by effigies** [reply]
Sorry, just a note that the slope and inter are also attributes on the dataobj. Will edit...
